Muon+JetMET part of the re-miniAOD fix

gpetruc opened the pull request:
PR that extends #17296 implementing the muon + jetmet side of the fixes.

The code can be activated, together with the E/gamma one, with 
```
from PhysicsTools.PatAlgos.slimming.customizeMiniAOD_MuEGFixMoriond2017 import customizeAll
process = customizeAll(process)
```

All the core functionality is there, but we may have some updates later to try reduce the amount of modules that are run (now there's quite some duplication) and understand the performance.
We put the PR anyway today so that the review of most of the code can already start before the sun sets on the west coast.

An AOD input file for testing with about 15 events with bad muons and 15 with gain switches is in /afs/cern.ch/work/g/gpetrucc/micro/CMSSW_8_0_25/src/horrorCollection.root

@mmarionncern @rafaellopesdesa @Sam-Harper @arizzi

Sam-Harper committed:
first implimentation of producer to make a collection from multifit with gs swapped with weights

Sam-Harper committed:
updates

Sam-Harper committed:
first version which makes the new SC (without widths), not yet in event

Sam-Harper committed:
now producing the superclusters

Sam-Harper committed:
swapping to PFCluster inputs so PFwidth can be computed

Sam-Harper committed:
add map maker so we can easily get the new superclusters when we have electrons and photons

Sam-Harper committed:
typo in the header guard fixed

Sam-Harper committed:
initial module to remake the electron cores

Sam-Harper committed:
first running version making new electrons

Sam-Harper committed:
removing debugging for now

Sam-Harper committed:
first version which remakes the electrons

Sam-Harper committed:
setting event content for the regresssion

Sam-Harper committed:
setting event content for the regresssion

Sam-Harper committed:
add a todo warning

Sam-Harper committed:
now getting the showershapes done

Sam-Harper committed:
new class to handle the common tools for gain switch stuff

Sam-Harper committed:
updating to push the showershape tools in gainswitch tools

Sam-Harper committed:
adding EB explicit checks

Sam-Harper committed:
more fixes and functions

Sam-Harper committed:
adding in null ref protection, (parents sc can be null)

Sam-Harper committed:
refactoring functions in to gainswitchtools

Sam-Harper committed:
simple ele and photon fixers

Sam-Harper committed:
typo fix

Sam-Harper committed:
remove debug

Sam-Harper committed:
first example config

Sam-Harper committed:
fixed debug message

Sam-Harper committed:
updating config

Sam-Harper committed:
cleaning up the config and renaming objects to be identified as GS fixed

Sam-Harper committed:
simple fixes

Sam-Harper committed:
adding egamma gs fix sequences

Sam-Harper committed:
example sequences for gains switch fix

Sam-Harper committed:
updating example input

Sam-Harper committed:
adding more ecal rec hit collections

gpetruc committed:
Tagger modules for bad global muons in MiniAOD and AOD

gpetruc committed:
FWLite version of it

argiro committed:
Fix for missing flagging of gain switch

rafaellopesdesa committed:
implement fix for photons

Sam-Harper committed:
correcting the H/E variables

rafaellopesdesa committed:
Merged GSIssueFix from repository Sam-Harper

Sam-Harper committed:
now correctly adds the sub-clusters of the refined superclusters to the event

gpetruc committed:
Also catch bad tracker-only muons

rafaellopesdesa committed:
port changes to PAT sequence through reducedEgamma object

gpetruc committed:
Added an option to always accept all events

Sam-Harper committed:
Merge pull request #2 from rafaellopesdesa/Sam-Harper/GSIssueFix

thanks for getting this sorted

rafaellopesdesa committed:
save a copy of modified electrons and photons in the miniAOD event

rafaellopesdesa committed:
keep the modified objects

Sam-Harper committed:
Merge pull request #3 from rafaellopesdesa/Sam-Harper/GSIssueFix

save a copy of modified electrons and photons in the miniAOD event

Sam-Harper committed:
making old products skip the current process

Sam-Harper committed:
bug fix, seed was not in the correct collection which was causing null matches

Sam-Harper committed:
adding customisation tools

Sam-Harper committed:
Merge branch 'Sam-Harper/GSIssueFixTest2' into Sam-Harper/GSIssueFix

rafaellopesdesa committed:
Merged GSIssueFix from repository Sam-Harper

rafaellopesdesa committed:
revert

rafaellopesdesa committed:
modification without touching the standard code

Sam-Harper committed:
fixing bug in showershape calculations

Sam-Harper committed:
fixing bug where the same cluster could be assigned to two superclusters

rafaellopesdesa committed:
Merge pull request #1 from Sam-Harper/GSIssueFixPATcustomize

Bug Fixes for electrons, photons and superclusters

Sam-Harper committed:
setting the SC matching to required the seed to be in the 3x3 of the orginal seed crystal

rafaellopesdesa committed:
Merge pull request #2 from Sam-Harper/GSIssueFixPATcustomize

tighting the seed crystal matching

Sam-Harper committed:
updating to save the org e/gamma

Sam-Harper committed:
fixing event content to avoid storing eles/phos twice (once as redo and once as pat

rafaellopesdesa committed:
Merge pull request #3 from Sam-Harper/GSIssueFixPATcustomize

customise function to save the orginal e/gamma objects

rafaellopesdesa committed:
Merged GSIssueFixPATcustomize from repository rafaellopesdesa

rafaellopesdesa committed:
save ECAL digis

rafaellopesdesa committed:
add ECAL digis only in thic cycle

rafaellopesdesa committed:
remove ECAL digis

rafaellopesdesa committed:
add them again... ECAL digis

gpetruc committed:
Fix default value of verbosity

gpetruc committed:
Don't import modules that we don't want to run

gpetruc committed:
Move every load inside customizeGSFixForPAT

rafaellopesdesa committed:
Merge pull request #4 from gpetruc/small_17296_cleanup

Refactor to have all cfg changes within customizeGSFixForPAT

rafaellopesdesa committed:
Merged GSIssueFixPATcustomize from repository rafaellopesdesa

rafaellopesdesa committed:
check is the fixed supercluster exists to avoid crashing

cmsbuild commented:
A new Pull Request was created by  @gpetruc (Giovanni Petrucciani) for CMSSW_8_0_X.

It involves the following packages:

CommonTools/ParticleFlow
DataFormats/MuonReco
DataFormats/ParticleFlowCandidate
PhysicsTools/PatAlgos
PhysicsTools/PatUtils
RecoEcal/EgammaClusterProducers
RecoEgamma/EgammaElectronProducers
RecoEgamma/EgammaIsolationAlgos
RecoEgamma/EgammaPhotonProducers
RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification
RecoLocalCalo/EcalRecProducers
RecoMET/METFilters
RecoParticleFlow/PFProducer


@cmsbuild, @cvuosalo, @slava77, @monttj, @davidlange6 can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @abbiendi, @rappoccio, @argiro, @mmarionncern, @imarches, @ahinzmann, @acaudron, @jhgoh, @lgray, @jdolen, @JyothsnaKomaragiri, @ferencek, @trocino, @Sam-Harper, @rafaellopesdesa, @nhanvtran, @gkasieczka, @schoef, @amagitte, @HuguesBrun, @battibass, @calderona, @mverzett, @cbernet, @mariadalfonso, @pvmulder, @bachtis this is something you requested to watch as well.
@davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

slava77 commented:
@gpetruc 
is it practical to have these changes factorized?
Except for maybe a file or two the changes are decoupled.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: fb0244914d45665ca711749b7c9e3069abbd1ede
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17476/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/fb0244914d45665ca711749b7c9e3069abbd1ede
https://github.com/cms-sw/cmssw/commit/e4825d0ce5c637acf98693234cf5bbed5fa914f4
https://github.com/cms-sw/cmssw/commit/9ef2fde5052eabcfb65b1fec3b70dd145ed37da8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17476/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17476/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17476/summary.html

rafaellopesdesa committed:
delete spurios files, add flag for users, and first code review modifications

rafaellopesdesa committed:
Merged GSIssueFixPATcustomize from repository rafaellopesdesa

slava77devel committed:
put unscheduled customs in a proper order: external methods are to be applied after all internal (miniAOD) are done

slava77 commented:
@gpetruc 
please pickup 
https://github.com/cms-sw/cmssw/commit/bd8c794491bb6365ba1971446f6ad662ee93cbc6
to be able to call the customization from cmsDriver.
AFAIK, we'd not be able to submit centrally without cmsDriver-able option.

rafaellopesdesa committed:
code review, part 2

slava77 commented:
apparently there is no magic to make this PR to update with the edits in #17296 
So, this PR will need to be rebased with changes there or wait until #17296 is merged and then rebased to get rid of all changes from #17296

slava77 commented on RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.cc:
```diff
@@ -9,8 +9,10 @@
 #include "CondFormats/DataRecord/interface/EcalWeightXtalGroupsRcd.h"
 #include "CondFormats/DataRecord/interface/EcalTBWeightsRcd.h"
 
-#include <FWCore/ParameterSet/interface/ConfigurationDescriptions.h>
```
edits in this file are outdated (different from what went to the release; before the PR review).
Merging locally suggests that there are no merge conflicts and a correct version is picked up.
Better clean up though

slava77 commented:
All my test jobs have crashed in 

```
#5  0x00007f48607a66e6 in edm::RefCoreWithIndex::operator=(edm::RefCoreWithIndex const&) () from libDataFormatsCommon.so
#6  0x00007f481eb8d3ff in EGRefinedSCFixer::produce(edm::Event&, edm::EventSetup const&) () from pluginR
ecoEcalEgammaClusterProducers.so
```
I suppose, this is because of some fixes missing in this PR but present in #17296

gpetruc commented:
Okay will rebase on #17296 later this morning

Il 30 Gen 2017 04:46, "Slava Krutelyov" <notifications@github.com> ha
scritto:

All my test jobs have crashed in

#5  0x00007f48607a66e6 in
edm::RefCoreWithIndex::operator=(edm::RefCoreWithIndex const&) () from
libDataFormatsCommon.so
#6  0x00007f481eb8d3ff in EGRefinedSCFixer::produce(edm::Event&,
edm::EventSetup const&) () from pluginR
ecoEcalEgammaClusterProducers.so

I suppose, this is because of some fixes missing in this PR but present in
#17296 <https://github.com/cms-sw/cmssw/pull/17296>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cms-sw/cmssw/pull/17308#issuecomment-275974721>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AEbbR12dLaFOunBjW187_Er84-Kh8NBbks5rXV0NgaJpZM4LwA6Y>
.

slava77 commented on CommonTools/ParticleFlow/plugins/PFMuonUntagger.cc:
```diff
@@ -0,0 +1,121 @@
+            reco::CandidatePtr p = badmuons->ptrAt(j);
+            if (p.isNonnull() && p.id() == muons.id()) {
+                reco::Muon &mu = (*copy)[p.key()];
+                mu.setType(mu.type() & ~reco::Muon::PFMuon);
```
I suppose, we do not care about pfP4 and other PF-related values in this muon.
Or is it preserved to know what it was originally?

gpetruc replied:
yes, we keep all the rest of the muon content in order for people to be able to un-do the tagging just by doing setType again.

slava77 commented on PhysicsTools/PatAlgos/interface/PATUserDataMerger.h:
```diff
@@ -99,11 +100,21 @@ namespace pat {
 pat::PATUserDataMerger<ObjectType, Operation>::PATUserDataMerger(const edm::ParameterSet & iConfig, edm::ConsumesCollector & iC) :
-  userDataSrc_(iConfig.getParameter<std::vector<edm::InputTag> >("src"))
+  userDataSrc_(iConfig.getParameter<std::vector<edm::InputTag> >("src")),
+  labelPostfixesToStrip_(iConfig.existsAs<std::vector<std::string>>("labelPostfixesToStrip") ? iConfig.getParameter<std::vector<std::string>>("labelPostfixesToStrip") : std::vector<std::string>())
```
is this ```existsAs``` necessary? there is a fillDescription already.

gpetruc replied:
I never understood if fillDescriptions was enough - testing now after addressing the other comments

gpetruc replied:
looks like in this case fillDesciptions is not enough, maybe because the parent class doesn't percolate the description filling to this helper.

rafaellopesdesa committed:
deals with the duplication in conversion and code review part 3

slava77 commented on PhysicsTools/PatAlgos/plugins/PATMuonSlimmer.cc:
```diff
@@ -83,11 +93,13 @@ pat::PATMuonSlimmer::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
+        for (unsigned int ipfh = 0, npfh = pf_.size(); ipfh < npfh; ++ipfh) {
+            iEvent.getByToken(pf_[ipfh], pf);
+            iEvent.getByToken(pf2pc_[ipfh], pf2pc);
+            for (unsigned int i = 0, n = pf->size(); i < n; ++i) {
```
better deref the handles outside the loop

slava77 commented on PhysicsTools/PatAlgos/plugins/PackedPFCandidateRefMixer.cc:
```diff
@@ -0,0 +1,91 @@
+            std::cout << "oldpfRef: " << oldpfRef.id() << " / " << oldpfRef.key()<< std::endl;
+            std::cout << "newpfRef: " << newpfRef.id() << " / " << newpfRef.key()<< std::endl;
+            std::cout << "and I have " << pf2pcs.size() << " rekey maps." << std::endl;
+            for (const auto & pf2pc : pf2pcs) {
```
Note that "cout" is not thread safe and will produce garbled output if the code is running multithreaded.
Better change to "LogInfo" (not printed by MessageLogger by default) or "LogPrint" (printed by MessageLogger by default) to use thread safe version.
We typically allow the pattern ``` if (debug) cout << ... ```, but it should be understood that it has issues and will not allow debugging of a job running with multiple threads.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/customizeMiniAOD_MuEGFixMoriond2017.py:
```diff
@@ -0,0 +1,138 @@
+
+    #redo the miniAOD data customization for new JEC modules created during the backup process
+    from PhysicsTools.PatAlgos.slimming.miniAOD_tools import miniAOD_customizeData
+    miniAOD_customizeData(process)
```
Is there any harm from calling this twice on modules existing before this customizeAll is called? 
The standard ConfigBuilder piece already generates a call to miniAOD_customizeAllData which includes miniAOD_customizeData before this customize will be called.

mmarionncern replied:
@slava77 I did not noticed any configuration issue by calling it twice, but in case there is something we missed it can be replaced. The second miniAOD_customizeData is needed to ensure that residuals JECs are properly taken into account in the new METs.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/extraSlimmedMETs_MuEGFixMoriond2017.py:
```diff
@@ -0,0 +1,156 @@
+    process.slimmedMETsMuEGClean.tXYUncForT01Smear = cms.InputTag("patPFMetT0pcT1SmearTxyMuEGClean")
+    del process.slimmedMETsMuEGClean.caloMET
+
+    process.MINIAODoutput.outputCommands.extend(["keep *_slimmedMETsUncorrected_*_*",
```
I guess this should use addKeepStatement instead

the production system module names are not necessarily the same

slava77 commented on PhysicsTools/PatAlgos/python/slimming/extraSlimmedMETs_MuEGFixMoriond2017.py:
```diff
@@ -0,0 +1,156 @@
+                        cms.InputTag("puppiMuonCorrection") )
+                                           )
+
+    process.MINIAODoutput.outputCommands.extend(["keep *_puppiMETEGCor_*_*",
```
addKeepStatement

slava77 commented on PhysicsTools/PatUtils/python/tools/muonRecoMitigation.py:
```diff
@@ -0,0 +1,84 @@
+        if runOnMiniAOD:
+            from RecoMET.METFilters.badGlobalMuonTaggersMiniAOD_cff import badGlobalMuonTagger, cloneGlobalMuonTagger
+        else:
+            from RecoMET.METFilters.badGlobalMuonTaggersAOD_cff import badGlobalMuonTagger, cloneGlobalMuonTagger
```
the way these are loaded, better define different module names in the _cff files and then load appropriately named clones.
Else, there is a chance a wrong load elsewhere overrides the result

gpetruc replied:
To my understanding, `muonRecoMitigiation` is not actually used in the re-miniAOD setup.
I leave it to @mmarionncern to fix it for the purpose of MET recipes on existing (Mini)AOD.

cmsbuild commented:
Pull request #17308 was updated. @cvuosalo, @monttj, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

Sam-Harper committed:
changing gs rec-hits to global reco

gpetruc commented:
Updated addressing most comments fro Slava + cherry-picking the cmsDriver commit + merging in the head of #17296 (so that we get the fixes for the crashes, and testing can proceed). 
I will do the rebase to clean up the diff but only after the final fix to the EG PR is done wrt the other issues (and put the addKeepStatement to the EG part)

rafaellopesdesa committed:
Merge pull request #5 from Sam-Harper/GSIssueRecHitsToGlobal

changing the GS rec-hits reco from weights to global

Sam-Harper committed:
fixing up comments

Commit:
update muon correction for MET from miniAOD to miniAOD

Sam-Harper committed:
throwing on duplicate clusters now optional & tagged

Sam-Harper committed:
modules renamed so they start with their final name

Sam-Harper committed:
clean up

rafaellopesdesa committed:
Merge pull request #7 from Sam-Harper/GSIssueNoThrow

stopping throws on duplicate superclusters in the event & other debugging info

gpetruc commented:
Tested the PR on the 591 events with possibly bad muons from @zdemirag `/store/user/zdemirag/metscan/aod/DoubleMuon/crab_pickEvents/170127_135603/0000` without any error. Checked event by event that the output of the new objects is sensible (on a few dozen events), and that the backup objects match with what is produced from a clean 8_0_25 (on the first 300 events or so).

The miniAODs are available in `/afs/cern.ch/work/g/gpetrucc/micro/for26/CMSSW_8_0_25/src/horrorCollection_Zeynep_reminiAOD.root` and `/afs/cern.ch/work/g/gpetrucc/micro/CMSSW_8_0_25/src/horrorCollection_ZeynepClean.root` (made with a clean 8_0_25).

Also tested on ~20k events of DoubleMuon Run2016F, again without any error and with a running time of ~3.5s/event (measured on a random mix of lxbatch nodes, reading from xrootd/AAA - so, take it more a sanity check than as a benchmark)

gpetruc committed:
Merge branch 'Sam-Harper/PackedCandNoPuppi' into new_base

gpetruc committed:
Merge remote-tracking branch 'my-cmssw/badMuonFilters_80X_v2' into new_base

slava77 commented:
I was testing with  	67d5eca.
In DoubleMuon 2016E 277069:90 I see the CPU time goes up by about x4 and PFMuonUntagger:muonsCleaned has already gotten one event where it took 127 s to run.
Something is bad there.

gpetruc committed:
DataFormats/MuonReco: dictionary for edm::Association<reco::MuonCollection>

gpetruc committed:
CommonTools/ParticleFlow: modules to allow removing a muon from particle flow, and resetting all references

gpetruc committed:
Support for stripping postfixes in userData labels

gpetruc committed:
PatAlgos helpers for extracting a sequence of modules dependent on some products, and for adding entries to the event content without knowing the output module name

gpetruc committed:
Tools to re-miniAOD with a cleaned collection of PF candidates after removing bad muons, and correct MET for the gain switch

gpetruc committed:
Dictionary for edm::PtrVector<reco::PFCandidate>

gpetruc commented:
On Tue, Jan 31, 2017 at 10:53 AM, Slava Krutelyov <notifications@github.com>
wrote:

> I was testing with 67d5eca
> <https://github.com/cms-sw/cmssw/commit/67d5eca3af1f68b8536a48d5738635b9bbbdda81>
> .
> In DoubleMuon 2016E 277069:90 I see the CPU time goes up by about x4 and
> PFMuonUntagger:muonsCleaned has already gotten one event where it took 127
> s to run.
>

Compared to plain miniAOD production, we re-run b-tagging (normally we only
re-run a few modules) + re-make ak4 & ak4 CHS & ak8 CHS jets twice + run
puppi twice (instead of just once) + run puppi substructure twice (instead
of once) + possibly compute 2 or 4 sets of met uncertainties (instead of 1).

Wrt PFMuonUntagger, the code is
https://github.com/gpetruc/cmssw/blob/for_80X_reminiAOD_integration/CommonTools/ParticleFlow/plugins/PFMuonUntagger.cc
and it doesn't seem like it can take 127s.
Maybe in unscheduled execution it gets assigned some bogus time? (e.g. the
I/O of reading the event)

Something is bad there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17308#issuecomment-276320428>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEbbR7DUrE7wHsD5WdxD9e0Wp4c0BGFtks5rXwSWgaJpZM4LwA6Y>
> .
>

arizzi commented:
@slava77  do you know the event number taking 127s ?

arizzi commented:
the bug is also in local, giovanni found it

gpetruc committed:
Fix bug in FMuonUntagger

slava77 commented:
it could indeed be some stalls.
Most events take ~0 time in this module, but then there is occasional (every ~10 or so) where it takes 50 -120 s.

arizzi commented:
with the fix we are back to about 1.8s/event (local file) and no more events stalling for 100s

gpetruc commented:
Fixed in
https://github.com/cms-sw/cmssw/commit/ce5fa91bed4be2220d6d873c3caa43179c89ca5e
Still testing the rebased branch before updating the PR.

On Tue, Jan 31, 2017 at 11:49 AM, Slava Krutelyov <notifications@github.com>
wrote:

> it could indeed be some stalls.
> Most events take ~0 time in this module, but then there is occasional
> (every ~10 or so) where it takes 50 -120 s.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17308#issuecomment-276332004>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEbbRxmpQ65nGyHjqomk_z1WMBKla05dks5rXxHDgaJpZM4LwA6Y>
> .
>

cmsbuild commented:
Pull request #17308 was updated. @cvuosalo, @monttj, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

gpetruc commented:
Included the bugfix + fix from matthieu on the miniAOD-from-miniAOD workflow, and rebased (with some squashing because it was too complicated to do otherwise).
The performance bugfix also fixes the handling of events with multiple bad muons, of which there were none in the 30 event test file, but there were a few in the ~591 events from Zeynep.

The cmsDriver option is `--customise_unsch PhysicsTools/PatAlgos/slimming/customizeMiniAOD_MuEGFixMoriond2017.customizeAll`

Pre-rebase version including all fixes is in `gpetruc:for_80X_reminiAOD_integration_before_rebase`

Checked that the new version gives identical results to the old one on both the 30 and the 591 events, for both nominal and backup collections.

Running some larger scale tests now, but that will take about an hour or so to complete (1k events/job).

Do you want me to also change https://github.com/gpetruc/cmssw/blob/for_80X_reminiAOD_integration/RecoEgamma/EgammaTools/python/egammaGainSwitchFixToolsForPAT_cff.py#L96-L106 to use `addKeepStatement` in this PR?

slava77 commented:
``` to use addKeepStatement in this PR? ```
yes, it will be more safe
(but will be invisible in local tests with cmsDriver)

gpetruc committed:
Use addKeepStatement in gammaGainSwitchFixToolsForPAT

cmsbuild commented:
Pull request #17308 was updated. @cvuosalo, @monttj, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

gpetruc commented:
done the addKeepStatement. 
Also, completed running on 19k events, with now an average time/event of ~1.8s.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 30e49964fe0e53ee672f431a480d8302cb487b6f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17538/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/5216918e2365d580401eff8ef20f7817c108dd8d
https://github.com/cms-sw/cmssw/commit/4cf92c0dcbb81097a2501a05161694f19a71081e
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17538/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17538/git-merge-result

cmsbuild commented:
Comparison job queued.

slava77 commented:
I'm waiting for comparisons ...

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17538/summary.html

slava77 commented:
.. not a +1 yet because there are changes in the default setup

for https://github.com/cms-sw/cmssw/pull/17308	30e4996
- code changes appear to be in line with the goals of the re-miniAOD; the defaults are unchanged and the re-miniAOD feature can be enabled with ``` --customise_unsch PhysicsTools/PatAlgos/slimming/customizeMiniAOD_MuEGFixMoriond2017.customizeAll ```
- jenkins test pass and comparisons with baseline show difference in miniAOD MET plots: 
<img width="840" alt="wf25202_slimmedmet_ele" src="https://cloud.githubusercontent.com/assets/4676718/22483865/4e4e477e-e7ff-11e6-998e-cfa11e1f3a9b.png">

- local tests were done with data in 2016E DoubleMuon DoubleEG and MET (run 277069 LS~90) and customizeAll applied
    - comparisons show only small changes expected for this inclusive sample; relying on the correctness of specific features to have been tested by developers and experts related to this PR
    - there are 425 more modules running in miniAOD and 8 have been removed (for a total of 1001, which includes miniAOD DQM in my tests)
    - single-thread job memory peak is up by 330MB (to 1.7GB), which looks stable after ~500 events
    - processing time is up by 50% , essentially all increase is from the Muon+JETMET parts of this PR (the ECAL part adds only ~5%)
    - event size is up, by more in egamma-heavy events; the Muon+JETMET increase in event content is only 5%

@gpetruc @mmarionncern 
please specify what is changing in the defaults
given that diffs showed up in MC as well, it's technically a problem

slava77 commented:
looking at default configs (without the customizeAll)

only baseline has (abbreviated list follows)
```
message: "adding L2L3Residual JEC for: patJetCorrFactorsNoHF"
process.Njettiness = cms.EDProducer("NjettinessAdder"
process.ak4PFJetsCHSNoHF
process.ak8PFJetsCHSFilteredMass
process.ak8PFJetsCHSTrimmedMass
process.patJetCorrFactorsNoHF
process.patJetFlavourAssociationLegacyNoHF
process.patJetFlavourAssociationNoHF
process.patJetGenJetMatchNoHF
process.patJetPartonAssociationLegacyNoHF
process.patJetPartonMatchNoHF
process.patJetPartonsLegacyNoHF
process.patJetPartonsNoHF
process.patJetsNoHF
process.pfCHSNoHF
process.pfNoPileUpJMENoHF
process.pfPileUpJMENoHF
process.tmpPFCandCollPtr
process.goodOfflinePrimaryVerticesNoHF
process.selectedPatJetsNoHF

```

Of the above, the following are actually not use
```
ak4PFJetsCHSNoHF
patJetCorrFactorsNoHF
patJetsNoHF
pfCHSNoHF
```

we have changes in parameters
```
process.basicJetsForMetNoHF
	src = cms.InputTag("patJetsNoHF")	=>	"patJets"
process.corrPfMetType1NoHF
	src = cms.InputTag("ak4PFJetsCHSNoHF")	=> "ak4PFJetsCHS"
```

hope this helps

slava77 commented:
Should we expect a fix to the NoHF MET imminently?

changes in the default configuration
```
process.basicJetsForMetNoHF
	src = cms.InputTag("patJetsNoHF")	=>	"patJets"
process.corrPfMetType1NoHF
	src = cms.InputTag("ak4PFJetsCHSNoHF")	=> "ak4PFJetsCHS"
```
don't look right

mmarionncern commented:
@slava77 Yes, it is a 1 line fix to enable again the jet reclustering which not automatic anymore when one recomputes the MET from pfCandidates in the MET tool. Currently making the commit

Commit:
fix NoHF MET

mmarionncern commented:
@slava77, @gpetruc 
the fix is here 
https://github.com/gpetruc/cmssw/pull/11

gpetruc committed:
Merge pull request #11 from mmarionncern/fix_NoHFMET

Re-enable the jet clustering for the noHF MET

cmsbuild commented:
Pull request #17308 was updated. @cvuosalo, @monttj, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

gpetruc commented:
testing

On Wed, Feb 1, 2017 at 8:45 AM, mmarionncern <notifications@github.com>
wrote:

> @slava77 <https://github.com/slava77>, @gpetruc
> <https://github.com/gpetruc>
> the fix is here
> gpetruc#11 <https://github.com/gpetruc/cmssw/pull/11>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17308#issuecomment-276592884>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEbbRwbpOzWJWIw9GpA_ONX4BRmeTgjhks5rYDgUgaJpZM4LwA6Y>
> .
>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: efba37c6c27b47e5716702fcf127a7fd84676206
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17555/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/efba37c6c27b47e5716702fcf127a7fd84676206
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17555/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17555/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17555/summary.html

slava77 commented:
<img width="864" alt="wf136 731_slimmedmetnohf" src="https://cloud.githubusercontent.com/assets/4676718/22505367/704c296e-e87b-11e6-97f1-7395c1be29e9.png">

still differences

slava77 commented:
in fact, I'm getting the same default PAT expanded config with efba37c and with 30e4996.

Commit:
fix reclusterJets parameter initialization

mmarionncern commented:
@slava77 @gpetruc 
Corrected in 
https://github.com/gpetruc/cmssw/pull/12
An initialization was not done properly

gpetruc committed:
Merge pull request #12 from mmarionncern/fix_NoHFMET

Fix noHF MET, second step

cmsbuild commented:
Pull request #17308 was updated. @cvuosalo, @monttj, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
expanded configs look promising

cmsbuild commented:
+1
Tested at: 4d1fca5cbd20cd6b78f14fb11b04d2f0c9cb5df5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17563/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/4d1fca5cbd20cd6b78f14fb11b04d2f0c9cb5df5
https://github.com/cms-sw/cmssw/commit/efba37c6c27b47e5716702fcf127a7fd84676206
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17563/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17563/git-merge-result

cmsbuild commented:
Comparison job queued.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/17308	 	4d1fca5

adding to comments in https://github.com/cms-sw/cmssw/pull/17308#issuecomment-276488241, noting changes since ce5fa91
- fixes were added to recover consistent picture in NoHF MET computation in miniAOD
- jenkins tests passed .. but comparisons are not out yet (the job haven't started yet an hour after being queued)
- local tests (2016E data PDs, 2K events) with the customizeAll added show that the CPU time has increase by ~50ms more ( 900 ms/ev from the total of this PR) and there are some changes in the NoHF MET  since the version in ce5fa91 [assuming that's correct]
- check of expanded config of the default PAT setup shows consistent configuration between the baseline and this PR with some changes in names of the modules related to NoHF MET and jet
- local tests with default PAT config show that there are no more changes in NoHF MET plots

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17308/17563/summary.html

slava77 commented:
looking at jenkins comparisons, it looks like my local test observations are confirmed and there are not changes in the (monitored) defaults
